src/script/hydra-notify

```diff
@@ -3,24 +3,91 @@
 use strict;
 use utf8;
 use Getopt::Long;
+use HTTP::Server::PSGI;
 use Hydra::Event;
 use Hydra::Event::BuildFinished;
 use Hydra::Helper::AddBuilds;
 use Hydra::Helper::Nix;
 use Hydra::Plugin;
 use Hydra::PostgresListener;
+use Parallel::ForkManager;
+use Prometheus::Tiny::Shared;
+use Time::HiRes qw( gettimeofday tv_interval );
 
 STDERR->autoflush(1);
 STDOUT->autoflush(1);
 binmode STDERR, ":encoding(utf8)";
 
+my $config = getHydraConfig();
+
+my $prom = Prometheus::Tiny::Shared->new;
+# Note: It is very important to pre-declare any metrics before using them.
+# Add a new declaration for any new metrics you create. See:
+# https://metacpan.org/pod/Prometheus::Tiny#declare
+$prom->declare(
+    "notify_plugin_executions",
+    type => "counter",
+    help => "Number of times each plugin has been called by channel."
+);
+$prom->declare(
+    "notify_plugin_runtime",
+    type => "histogram",
+    help => "Number of seconds spent executing each plugin by channel."
+);
+$prom->declare(
+    "notify_plugin_success",
+    type => "counter",
+    help => "Number of successful executions of this plugin on this channel."
+);
+$prom->declare(
+    "notify_plugin_error",
+    type => "counter",
+    help => "Number of failed executions of this plugin on this channel."
+);
+$prom->declare(
+    "event_loop_iterations",
+    type => "counter",
+    help => "Number of iterations through the event loop. Incremented at the start of the event loop."
+);
+$prom->declare(
+    "event_received",
+    type => "counter",
+    help => "Timestamp of the last time a new event was received."
+);
+$prom->declare(
+    "notify_event",
+    type => "counter",
+    help => "Number of events received on the given channel."
+);
+$prom->declare(
+    "notify_event_error",
+    type => "counter",
+    help => "Number of events received that were unprocessable by channel."
+);
+
+my $promCfg = Hydra::Helper::Nix::getHydraNotifyPrometheusConfig($config);
+if (defined($promCfg)) {
+    print STDERR "Starting the Prometheus exporter, listening on http://${\$promCfg->{'listen_address'}}:${\$promCfg->{'port'}}/metrics.\n";
+    my $fork_manager = Parallel::ForkManager->new(1);
+    $fork_manager->start_child("metrics_exporter", sub {
+        my $server = HTTP::Server::PSGI->new(
+            host => $promCfg->{"listen_address"},
+            port => $promCfg->{"port"},
+            timeout => 1,
+        );
+
+        $server->run($prom->psgi);
+    });
```

Comment:
This is a pretty ugly thing. In order to export metrics we basically need a background thread / process serving the data. Perl's threads are "do not use" so here we are. We're using Prometheus::Tiny::Shared which uses a mmap'd file to share the state from the event loop process with the exporter process. This is doubly not great as every write is a json encode, and every read is a json decode. However, again, here we are.

Reply:
if this is as gross as you describe: Couldn't be write a text file with the metrics instead? Any webserver is able to serve that and the prometheus node exporter is able to pick up text files.

Reply:
Good question. `Net::Prometheus` (an objectively better Prometheus implementation) supports `->render`ing to text, which we could write out to a file.

However:

1. Writing to a text file means we don't get valid `up` metrics: if hydra-notify crashes we're still getting metrics as if things are fine. This could be resolved by having the second process serve it.
2. The metrics in the file are only as up to date as the most recent write. If we don't write out to the file on every metric update, we lose valuable hints as to what is going on.

I didn't see a way to have Net::Prometheus write out on every update, but perhaps that could be done?

What do you think?

Reply:
Regarding the 1. point: We could always add a timestamp to the file.
2. How many writes / second are we talking here? Writing out 9 metrics sounds relatively cheap but we will have to do it atomically (so new file and then move it).

To make that work we could probably wrap our usage of the library such that it writes to a file whenever we set a metric? Another option would be to update the file after each time we try to execute a notification?